### ISSUE TITLE
Thread Blocking in Magnetometer driver

### ISSUE THREAD

**Issue by mazahner** [PX4/Firmware#5081]
Hello

I'm using the hmc5883 driver in Firmware/src/drivers/hmc5883 off the latest stable branch.
It works fine on the i2c device. But, When i'm trying to set the calibration in the driver, my thread blocks:

In my application:

``` cpp
  int mag_fd = open(MAG0_DEVICE_PATH,0);
  if (mag_fd <= 0) {
    printf("couldn't open %s", MAG0_DEVICE_PATH);
  } else {
    mag_scale magscale;
   /* fill magscale*/
    if (OK != ioctl(mag_fd ,MAGIOCSSCALE,(unsigned long) (&magscale))) {
      printf("failed setting calib %s", MAG0_DEVICE_PATH);;
    }
    close(mag_fd);
  }
```

Thread blocks at ioctl();

I managed to make the ioctl() call work by commenting out these two lines in hmc5883.cpp:

``` cpp
int HMC5883::check_calibration()
{
    bool offset_valid = (check_offset() == OK);
    bool scale_valid  = (check_scale() == OK);

    if (_calibrated != (offset_valid && scale_valid)) {
    ---->   //warnx("mag cal status changed %s%s", (scale_valid) ? "" : "scale invalid ",
    ---->   //(offset_valid) ? "" : "offset invalid");
        _calibrated = (offset_valid && scale_valid);
    }

    /* return 0 if calibrated, 1 else */
    return (!_calibrated);
}
```

Any Idea why this blocks my caller thread?

Thanks for you help, 
Markus

**Comment by julianoes** [comment]
Strange, so you are saying that the ioctl never returns?

Can you retest after with master which includes this fix: #5062.

**Comment by bkueng** [comment]
Possibly a stack overflow. Can you look at the stack usage using `top` on the nsh console?

**Comment by mazahner** [comment]
Thanks for you're replies!
from the top i can tell that it blocks the thread. (it does not segfault which would happen on a stackoverflow)
i see the specific thread that was called with the flag waiting on semaphore.

@julianoes i'll try with the mater branch.

**Comment by dagar** [comment]
@mazahner any update here? Did you try master?

**Comment by LorenzMeier** [comment]
This was anyway wrong on that level. Drivers should never print to the console. I fixed that in 256222d44d.
